syntax = "proto3"; // 声明了protobuf的版本

package fixbug;  // 声明了代码所在的包（对于c++来说是namespace)

// 定义下面的选项，表示生成service服务类和rpc方法描述，默认不生成
option cc_generic_services = true;
// 数据  列表  映射表

message ResultCode{
    int32  errcode = 1;
    bytes errmsg = 2;
}


// protoc test.proto --cpp_out=./
// 定义登录请求消息类型  name pwd
message LoginRequest{
//    string name =1; // 定义字段序号
//    string pwd  = 2;
    bytes name =1; // 定义字段序号
    bytes pwd  = 2;
    map<int32 ,string >test=3;
}


// 定义登录响应消息类型
message LoginResponse{
    ResultCode result =1; // 对象类型
    bool   success=2;
}

message GetFriendListRequest{
    uint32 userid =1;
}


message User{
    bytes name=1;
    uint32 age=2;
    enum Sex{
        MAN =0;
        WOMAN =1;
    }
    Sex sex = 3;
}


message GetFriendListResponse{
    ResultCode result =1;
    // repeated 重复使用  User类有个列表
    repeated User friend_list = 2;  // 定义了一个列表类型
}

// 在protobuf里面怎么定义描述rpc方法的类型 --service
service UserServiceRpc{
    // RPC（Remote Procedure Call）并不是一种数据类型，而是一种通信模式或协议，它可以在不同的系统之间进行远程调用。
    // "rpc" 是用来声明一个远程过程调用（RPC）的关键字。
    // 它指示了在定义的 Protocol Buffers 文件中存在一个 RPC 方法。这个声明告诉编译器生成与该 RPC 方法相关的代码，以便在客户端和服务器端进行使用。
    rpc Login(LoginRequest) returns (LoginResponse);
    rpc GetFriendLists(GetFriendListRequest) returns (GetFriendListResponse);
}